File: frontend/src/lib/mapEmbed.ts
/** Accepts maps embed URL or full <iframe ...> HTML. Rejects plain text. */
export function resolveMapEmbed(raw?: string | null): string | null {
  const v = (raw ?? '').trim()
  if (!v) return null

  const fromIframe = v.match(/<iframe[^>]+src=["']([^"']+)["']/i)?.[1]?.trim()
  const candidate = (fromIframe || v).trim()

  // Sentences / pasted body text are not map URLs
  if (!fromIframe && (/\s/.test(candidate) || candidate.length > 500)) return null

  try {
    const u = new URL(candidate)
    if (u.protocol !== 'http:' && u.protocol !== 'https:') return null
    const host = u.hostname.toLowerCase()
    const href = u.toString().toLowerCase()
    const looksLikeMap =
      host.includes('google.') ||
      host.includes('googleapis.com') ||
      host.includes('openstreetmap.org') ||
      host.includes('osm.org') ||
      host.includes('yandex.') ||
      host.includes('mapy.cz') ||
      host.includes('maps.apple.com') ||
      href.includes('/maps') ||
      href.includes('map=') ||
      href.includes('embed')
    return looksLikeMap ? u.toString() : null
  } catch {
    return null
  }
}

File: backend/internal/service/guide/mapembed.go
package guide

import (
	"net/url"
	"regexp"
	"strings"
)

var iframeSrcRe = regexp.MustCompile(`(?i)<iframe[^>]+src=["']([^"']+)["']`)

// Exact host allowlist for map embeds (https only).
var mapHosts = map[string]struct{}{
	"www.google.com":        {},
	"maps.google.com":       {},
	"maps.googleapis.com":   {},
	"www.google.com.ua":     {},
	"www.openstreetmap.org": {},
	"openstreetmap.org":     {},
	"www.osm.org":           {},
	"osm.org":               {},
	"yandex.ru":             {},
	"yandex.ua":             {},
	"yandex.com":            {},
	"www.yandex.ru":         {},
	"www.yandex.ua":         {},
	"www.yandex.com":        {},
	"mapy.cz":               {},
	"www.mapy.cz":           {},
	"maps.apple.com":        {},
}

// ResolveMapEmbed accepts a maps embed URL or full <iframe src="..."> HTML.
// Plain text (e.g. pasted program paragraphs) is rejected.
func ResolveMapEmbed(raw string) string {
	v := strings.TrimSpace(raw)
	if v == "" {
		return ""
	}

	fromIframe := ""
	if m := iframeSrcRe.FindStringSubmatch(v); len(m) > 1 {
		fromIframe = strings.TrimSpace(m[1])
	}
	candidate := v
	if fromIframe != "" {
		candidate = fromIframe
	} else if strings.ContainsAny(candidate, " \t\n\r") || len(candidate) > 500 {
		return ""
	}

	u, err := url.Parse(candidate)
	if err != nil || u.Scheme != "https" {
		return ""
	}
	host := strings.ToLower(u.Hostname())
	if _, ok := mapHosts[host]; !ok {
		return ""
	}
	return u.String()
}
